src/Crypto.gren

```diff
@@ -0,0 +1,2227 @@
+        bytes
+
+
+{-| Sign some `Bytes` with the HMAC algorithm. This produces a `Signature` (which 
+is just some `Bytes`). The `Signature` can be used with the cooresponding verification function 
+to verify that the passed `Bytes` were signed with the passed key.
+
+The returned `Task` should not fail. If it does, please file a ticket!
+-}
+signWithHmac : Key HmacKey HmacKeyParams -> Bytes -> Task x Signature
+signWithHmac (Key key _) bytes =
+    Gren.Kernel.Crypto.signWithHmac 
+        key 
+        bytes
+
+
+
+-- VERIFY
+
+
+{-| Verify that some `Bytes` were signed with the passed `Signature` with the 
+RSA-SSAPKCS1v1.5 algorithm.
+
+This function produces no values. Instead, the `Task` succeeds if the passed 
+signature is valid and fails otherwise.
+-}
+verifyWithRsaSsaPkcs1V1_5 : PublicKey RsaSsaPkcs1V1_5Key RsaKeyParams -> Signature -> Bytes -> Task x a
+verifyWithRsaSsaPkcs1V1_5 (PublicKey (Key key _)) signature bytes =
+    Gren.Kernel.Crypto.verifyWithRsaSsaPkcs1V1_5
+        key
+        signature
+        bytes
+
+
+{-| Verify that some `Bytes` were signed with the passed `Signature` with the 
+RSA-PSS algorithm.
+
+This function produces no values. Instead, the `Task` succeeds if the passed 
+signature is valid and fails otherwise.
+-}
+verifyWithRsaPss : RsaPssParams -> PublicKey RsaPssKey RsaKeyParams -> Signature -> Bytes -> Task x a
+verifyWithRsaPss { salt } (PublicKey (Key key _)) signature bytes =
+    Gren.Kernel.Crypto.verifyWithRsaPss
+        salt
+        key
+        signature
+        bytes
+
+
+{-| Verify that some `Bytes` were signed with the passed `Signature` with the 
+ECDSA algorithm.
+
+This function produces no values. Instead, the `Task` succeeds if the passed 
+signature is valid and fails otherwise.
+-}
+verifyWithEcdsa : DigestAlgorithm -> PublicKey EcdsaKey EcKeyParams -> Signature -> Bytes -> Task x a
+verifyWithEcdsa hash (PublicKey (Key key _)) signature bytes =
+    Gren.Kernel.Crypto.verifyWithEcdsa
+        (digestAlgorithmToString hash)
+        key
+        signature
+        bytes
+
+
+{-| Verify that some `Bytes` were signed with the passed `Signature` with the 
+HMAC algorithm.
+
+This function produces no values. Instead, the `Task` succeeds if the passed 
+signature is valid and fails otherwise.
+-}
+verifyWithHmac : Key HmacKey HmacKeyParams -> Signature -> Bytes -> Task x a
+verifyWithHmac (Key key _) signature bytes =
+    Gren.Kernel.Crypto.verifyWithHmac
+        key
+        signature
+        bytes
+
+
+
+-- DIGEST
+
+
+{-| Supported algorithms suitable for digesting data.
+
+Note: The algorithm `SHA1` is supported by the WebCrypto API, but not available 
+in this package due to known security vunerabilities.
+-}
+type DigestAlgorithm
+    = Sha256
+    | Sha384
+    | Sha512
+
+
+{-| Take some `Bytes` and create a hash from them using the passed `DigestAlgorithm`.
+This operation should always succeed.
+-}
+digest : DigestAlgorithm -> Bytes -> SecureContext -> Task x Bytes
```

Comment:
I believe it would be more convenient if `SecureContext` was the first parameter (this may apply to other functions as well).

If you imagine doing things in a pipeline, I'd imagine the `Bytes` being the piped value most often.

Reply:
@robinheghan I changed the order of arguments for the `digest` function. The other functions using `SeucreContext` are key import and generation functions which have signatures like `AesKeyParams -> SecureContext -> Task x (Key AesCtrKey AesKeyParams)`. This is how I'm using this in some tests (just an example, so this isn't 100% how it'd appear in a codebase):

```elm
getSecureContext
        |> Task.mapError (\_ -> Crypto.RsaKeyGenerationError)
        |> Task.andThen (generateRsaOaepKeyPair { ...params })
```

I can change the argument order for key generation / import functions, too, though! Let me know what you think.

Reply:
I think for NodeJS applications, you're likely to retrieve the `SecureContext` value as part of the application initialization. I guess for browsers, where we don't have the app initialization pattern then this might more sense 🤔 

Leave it the way it is now. 👍